{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import itertools\n",
    "from Bio import SeqIO\n",
    "\n",
    "\n",
    "class Action:\n",
    "    MATCH = 1\n",
    "    GAP_I = 2\n",
    "    GAP_J = 3\n",
    "    TWO_GAPS = 4\n",
    "\n",
    "\n",
    "class MultiSequenceAligner:\n",
    "\n",
    "    def __init__(self,\n",
    "                 match: int = 5,\n",
    "                 mismatch: int = -2,\n",
    "                 indel: int = -4,\n",
    "                 two_gaps: int = 0):\n",
    "        \"\"\"\n",
    "        :param match: Score if AA match\n",
    "        :param mismatch: Score if AA mismatch\n",
    "        :param indel: Linear gap penalty (applied for every gap)\n",
    "        :param two_gaps: Additional gap penalty applied if two or more sequences have a gap at given position\n",
    "        \"\"\"\n",
    "        self.match = match\n",
    "        self.mismatch = mismatch\n",
    "        self.indel = indel\n",
    "        self.two_gaps = two_gaps\n",
    "\n",
    "    def get_alignment_matrix(self, sequences: list[str], method: str = \"global\") -> tuple[np.ndarray, np.ndarray]:\n",
    "        \"\"\"\n",
    "        Returns a matrix of scores for all possible alignments of the given sequences and a backtrack matrix.\n",
    "        :param sequences: List of sequences to align\n",
    "        :param method: Method to use for alignment. Either \"global\" or \"local\"\n",
    "        :return: Matrix of scores for all possible alignments of the given sequences and a backtrack matrix\n",
    "        \"\"\"\n",
    "        assert method in [\"global\", \"local\"], \"Method must be either 'global' or 'local'\"\n",
    "\n",
    "        # Initialize matrix\n",
    "        # n-dimensional matrix where n is the number of sequences, each dimension is the length of the sequence + 1 (for the inital gap)\n",
    "        dimensions = [len(sequence) + 1 for sequence in sequences]\n",
    "        matrix = np.zeros(dimensions)\n",
    "\n",
    "        # Backtrack matrix contains taken direction to get the score\n",
    "        backtrack_dimensions = dimensions + [len(sequences)]\n",
    "        backtrack_matrix = np.zeros(backtrack_dimensions)\n",
    "\n",
    "        # Fill matrix\n",
    "        idx_value_pairs = np.ndenumerate(matrix)\n",
    "        # sort on sum of indices to ensure that the first value is the one with the lowest index sum\n",
    "        idx_value_pairs = sorted(idx_value_pairs, key=lambda x: sum(x[0]))\n",
    "        for idx, value in idx_value_pairs:\n",
    "            score, neighbour = self._get_score(idx, sequences, matrix, method)\n",
    "            matrix[idx] = score\n",
    "            backtrack_matrix[idx] = neighbour\n",
    "\n",
    "        return matrix, backtrack_matrix\n",
    "\n",
    "    def _get_score(self, idx: tuple[int], sequences: list[str], matrix: np.ndarray, method: str) -> tuple[\n",
    "        int, tuple[int]]:\n",
    "        \"\"\"\n",
    "        Returns the score for the given position in the matrix. The score is s completely analogous to the pairwise\n",
    "        case, only now the scores for each position are equal to the sum of the individual\n",
    "        pairwise comparisons (i.e. a position that is identical for three sequences has a\n",
    "        score of 5s1,s2 + 5s1,s3 + 5s2,s3 = 15).\n",
    "\n",
    "        :param idx: Index of the position in the matrix\n",
    "        :param sequences: List of sequences to align\n",
    "        :param matrix: Partially completed matrix of scores for all possible alignments of the given sequences\n",
    "        :param method: Method to use for alignment. Either \"global\" or \"local\"\n",
    "        :return: Score for given position in the matrix and the index of the neighbour that generated the score.\n",
    "        \"\"\"\n",
    "        # if idx all zeros, return 0\n",
    "        if all([i == 0 for i in idx]):\n",
    "            return 0, None\n",
    "\n",
    "        num_sequences = len(sequences)\n",
    "\n",
    "        neighbours = self._get_preceeding_neigbours(idx)\n",
    "\n",
    "        # get scores for all neighbours\n",
    "        scores = []\n",
    "        for neighbour in neighbours:\n",
    "\n",
    "            # if any of the indices is negative, return -inf\n",
    "            if any([i < 0 for i in neighbour]):\n",
    "                scores.append(-np.inf)\n",
    "                continue\n",
    "\n",
    "            pairs = self._get_pairs(num_sequences)\n",
    "            total = matrix[neighbour]\n",
    "            for i, j in pairs:\n",
    "                score, action = self._get_pairwise_score(neighbour, i, j, matrix, sequences, idx)\n",
    "                total += score\n",
    "            scores.append(total)\n",
    "\n",
    "        max_score = max(scores)\n",
    "        max_idx = scores.index(max_score)\n",
    "        max_score_neighbour = neighbours[max_idx]\n",
    "\n",
    "        if method == \"global\":\n",
    "            return max_score, max_score_neighbour\n",
    "        elif method == \"local\":\n",
    "            return max(max_score, 0), max_score_neighbour if max_score > 0 else None\n",
    "        else:\n",
    "            raise ValueError(\"Method must be either 'global' or 'local'\")\n",
    "\n",
    "    def _get_preceeding_neigbours(self, idx: tuple[int]) -> list[tuple[int]]:\n",
    "        \"\"\"\n",
    "        Returns a list of all possible preceeding neighbours of the given index. A neighbour is defined as a tuple of indices\n",
    "        where each index is either the same as the given index or one less.\n",
    "        :param idx: Index of the position in the matrix\n",
    "        :return: List of all possible preceeding neighbours of the given index\n",
    "        \"\"\"\n",
    "        num_sequences = len(idx)\n",
    "        all_neighbours = list(itertools.product([0, -1], repeat=num_sequences))\n",
    "        all_neighbours.remove(tuple([0 for _ in range(num_sequences)]))\n",
    "        neighbours = [tuple(np.array(idx) + np.array(neighbour)) for neighbour in all_neighbours]\n",
    "        return neighbours\n",
    "\n",
    "    def _get_pairwise_score(self, neighbour: tuple[int], i: int, j: int, matrix: np.ndarray, sequences: list[str],\n",
    "                            idx: tuple[int]) -> tuple[int, int]:\n",
    "        \"\"\"\"\n",
    "        Returns the score of a pair for given neigbour to idx transition\n",
    "        :param neighbour: Neighbour position\n",
    "        :param i: Index of first sequence\n",
    "        :param j: Index of second sequence\n",
    "        :param matrix: Partially completed matrix of scores for all possible alignments of the given sequences\n",
    "        :param sequences: List of sequences to align\n",
    "        :param idx: score for given pair and action that generated the score\n",
    "        \"\"\"\n",
    "        neighbour_i = neighbour[i]\n",
    "        neighbour_j = neighbour[j]\n",
    "\n",
    "        idx_i = idx[i]\n",
    "        idx_j = idx[j]\n",
    "\n",
    "        diff_i = idx_i - neighbour_i\n",
    "        diff_j = idx_j - neighbour_j\n",
    "\n",
    "        # determin the action based on diff_i and diff_j\n",
    "        if diff_i == 1 and diff_j == 1:\n",
    "            action = Action.MATCH\n",
    "        elif diff_i == 1 and diff_j == 0:\n",
    "            action = Action.GAP_J\n",
    "        elif diff_i == 0 and diff_j == 1:\n",
    "            action = Action.GAP_I\n",
    "        elif diff_i == 0 and diff_j == 0:\n",
    "            action = Action.TWO_GAPS\n",
    "        else:\n",
    "            raise ValueError(f\"Invalid offset: {diff_i}, {diff_j}\")\n",
    "\n",
    "        if action == Action.MATCH:\n",
    "            # match\n",
    "            aa_i = sequences[i][idx_i - 1]\n",
    "            aa_j = sequences[j][idx_j - 1]\n",
    "            score = self.match if aa_i == aa_j else self.mismatch\n",
    "\n",
    "        elif action == Action.GAP_I:\n",
    "            # gap in j\n",
    "            score = self.indel\n",
    "        elif action == Action.GAP_J:\n",
    "            # gap in i\n",
    "            score = self.indel\n",
    "        elif action == Action.TWO_GAPS:\n",
    "            # two gaps\n",
    "            score = self.two_gaps\n",
    "        else:\n",
    "            raise ValueError(\"Invalid offset\")\n",
    "\n",
    "        return score, action\n",
    "\n",
    "    def _get_pairs(self, n):\n",
    "        \"\"\"\n",
    "        Returns all possible pairs of n numbers (i.e. for n=3, returns [(0,1), (0,2), (1,2)])\n",
    "        :param n: Number of numbers\n",
    "        :return: List of all possible pairs of n numbers\n",
    "        \"\"\"\n",
    "        pairs = []\n",
    "        for i in range(n):\n",
    "            for j in range(i + 1, n):\n",
    "                pairs.append((i, j))\n",
    "        return pairs\n",
    "\n",
    "    def alignment(self, sequences: list[str], method: str):\n",
    "        \"\"\"\n",
    "        Returns the optimal alignment of the given sequences\n",
    "        :param sequences: List of sequences to align\n",
    "        :param method: Method to use for alignment, either \"global\" or \"local\"\n",
    "        :return: List of aligned sequences\n",
    "        \"\"\"\n",
    "        return self.global_alignment(sequences) if method == \"global\" else self.local_alignment(sequences)\n",
    "\n",
    "    def global_alignment(self, sequences: list[str]):\n",
    "        \"\"\"\n",
    "        Returns the optimal global alignment of the given sequences\n",
    "        :param sequences: List of sequences to align\n",
    "        :return: List of aligned sequences\n",
    "        \"\"\"\n",
    "        alignment_matrix, backtrack_matrix = self.get_alignment_matrix(sequences, method=\"global\")\n",
    "        return self._get_alignment(sequences, alignment_matrix, backtrack_matrix, method=\"global\")\n",
    "\n",
    "    def local_alignment(self, sequences: list[str]):\n",
    "        \"\"\"\n",
    "        Returns the optimal local alignment of the given sequences\n",
    "        :param sequences: List of sequences to align\n",
    "        :return: List of aligned sequences\n",
    "        \"\"\"\n",
    "        alignment_matrix, backtrack_matrix = self.get_alignment_matrix(sequences, method=\"local\")\n",
    "        return self._get_alignment(sequences, alignment_matrix, backtrack_matrix, method=\"local\")\n",
    "\n",
    "    def _get_alignment(self, sequences: list[str], alignment_matrix: np.ndarray, backtrack_matrix: np.ndarray,\n",
    "                       method: str):\n",
    "        \"\"\"\n",
    "        Backtraces through the alignment matrix to get the optimal alignment of the given sequences\n",
    "        :param sequences: List of sequences to align\n",
    "        :param alignment_matrix: Matrix of scores for all possible alignments of the given sequences\n",
    "        :param backtrach_matrix: Matrix of backtrach directions for all possible alignments of the given sequences\n",
    "        :param method: Method to use for alignment. Either \"global\" or \"local\"\n",
    "        :return: List of aligned sequences\n",
    "        \"\"\"\n",
    "        aligned_sequences_reversed = [\"\"] * len(sequences)\n",
    "\n",
    "        current_position = [len(sequence) for sequence in sequences]\n",
    "        if method == \"local\":\n",
    "            current_position = np.unravel_index(np.argmax(alignment_matrix), alignment_matrix.shape)\n",
    "\n",
    "        while True:\n",
    "\n",
    "            previous_neighbour = backtrack_matrix[tuple(current_position)]\n",
    "\n",
    "            for sequence_idx in range(len(sequences)):\n",
    "                if current_position[sequence_idx] == previous_neighbour[sequence_idx]:\n",
    "                    aligned_sequences_reversed[sequence_idx] += \".\"\n",
    "                else:\n",
    "                    aligned_sequences_reversed[sequence_idx] += sequences[sequence_idx][\n",
    "                        current_position[sequence_idx] - 1]\n",
    "\n",
    "            if self._backtrack_break_condition(previous_neighbour, method, alignment_matrix):\n",
    "                break\n",
    "\n",
    "            current_position = previous_neighbour\n",
    "\n",
    "            # change the floats (e.g. 2.000000) in current_position to ints\n",
    "            current_position = [int(i) for i in current_position]\n",
    "\n",
    "        aligned_sequences = [sequence[::-1] for sequence in aligned_sequences_reversed]\n",
    "        return aligned_sequences\n",
    "\n",
    "    def _backtrack_break_condition(self, previous_neighbour, method, alignment_matrix):\n",
    "        if method == \"global\":\n",
    "            return np.all(previous_neighbour == 0)\n",
    "        elif method == \"local\":\n",
    "            previous_neighbour = [int(i) for i in previous_neighbour]\n",
    "            return np.all(alignment_matrix[tuple(previous_neighbour)] == 0)\n",
    "        else:\n",
    "            raise ValueError(f\"Invalid method: {method}\")\n",
    "\n",
    "    def _get_match_score(self, aa_i: str, aa_j: str) -> float:\n",
    "        \"\"\"\n",
    "        Returns the score for matching the given amino acids (or other characters)\n",
    "        :param aa_i: Amino acid i\n",
    "        :param aa_j: Amino acid j\n",
    "        :return: Score for matching the given amino acids\n",
    "        \"\"\"\n",
    "        if aa_i is None or aa_j is None:\n",
    "            return -np.inf\n",
    "        return self.match if aa_i == aa_j else self.mismatch\n",
    "\n",
    "    def align_fasa(self, input_file: str, output_file: str = None, method: str = \"global\") -> None:\n",
    "        \"\"\"\n",
    "        Aligns the sequences in the given fasta file and writes the result to the given output file. If no output file is given, the result is printed to the console.\n",
    "        :param input_file: Path to the fasta file containing the sequences to align\n",
    "        :param output_file: Path to the output file to write the result to\n",
    "        :param method: Method to use for alignment. Either \"global\" or \"local\"\n",
    "        \"\"\"\n",
    "        sequence_dict: dict = self._read_fasta(input_file)\n",
    "\n",
    "        ids, sequences = zip(*sequence_dict.items())\n",
    "\n",
    "        aligned_sequences = self.alignment(sequences, method=method)\n",
    "\n",
    "        aligned_sequence_dict = dict(zip(ids, aligned_sequences))\n",
    "\n",
    "        output_str = \"\\n\".join([f\"{id}: {sequence}\" for id, sequence in aligned_sequence_dict.items()])\n",
    "\n",
    "        if output_file is None:\n",
    "            print(output_str)\n",
    "        else:\n",
    "            with open(output_file, \"w\") as f:\n",
    "                f.write(output_str)\n",
    "\n",
    "    def _read_fasta(self, input_file: str) -> dict[str, str]:\n",
    "        \"\"\"\n",
    "        Reads the given fasta file and returns a dictionary containing the ids and sequences\n",
    "        :param input_file: Path to the fasta file\n",
    "        :return: Dictionary containing the ids and sequences\n",
    "        \"\"\"\n",
    "        sequence_dict = {}\n",
    "        for record in SeqIO.parse(input_file, \"fasta\"):\n",
    "            sequence_dict[record.id] = str(record.seq)\n",
    "        return sequence_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "outputs": [],
   "source": [
    "def print_alignments(sequences, method='global', params=None):\n",
    "    if params is None:\n",
    "        params = {}\n",
    "    aligner = MultiSequenceAligner(**params)\n",
    "    alignments = aligner.alignment(sequences, method=method)\n",
    "    print('\\n'.join(alignments))"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACTGG.TC.A\n",
      ".CAGGGTC.A\n",
      "CCAGGGACCA\n"
     ]
    }
   ],
   "source": [
    "print_alignments([\"ACTGGTCA\", \"CAGGGTCA\", \"CCAGGGACCA\"])"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "outputs": [
    {
     "data": {
      "text/plain": "array([[[  0.,  -8., -16.],\n        [ -8.,  -3., -11.],\n        [-16., -11., -13.]],\n\n       [[ -8.,  -3., -11.],\n        [ -3.,  15.,   7.],\n        [-11.,   7.,   5.]],\n\n       [[-16., -11.,  -6.],\n        [-11.,   7.,  12.],\n        [-13.,   5.,  16.]],\n\n       [[-24., -19., -14.],\n        [-19.,  -1.,   4.],\n        [-14.,   4.,   9.]]])"
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aligner = MultiSequenceAligner()\n",
    "seqs = [\"ABC\", \"AC\", \"AB\"]\n",
    "alignment_matrix, backtrack_matrix = aligner.get_alignment_matrix(seqs, method=\"global\")\n",
    "alignment_matrix"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "outputs": [
    {
     "data": {
      "text/plain": "['ABC', 'A.C', 'AB.']"
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aligner.alignment(seqs, method=\"global\")"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CD\n",
      "CD\n"
     ]
    }
   ],
   "source": [
    "print_alignments([\"ABCDEF\", \"CD\"], method=\"local\")"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AB\n",
      "AC\n",
      "AB\n",
      ".B\n"
     ]
    }
   ],
   "source": [
    "print_alignments([\"ABC\", \"AC\", \"AB\", \"BCD\"], method=\"local\")"
   ],
   "metadata": {
    "collapsed": false
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
